pkg/sources/gcs/gcs_test.go

```diff
@@ -289,3 +294,57 @@ func TestSourceChunks_ListObjects_Error(t *testing.T) {
 	err := source.Chunks(ctx, chunksCh)
 	assert.True(t, err != nil)
 }
+
+func TestSourceChunks_ProgressSet(t *testing.T) {
+	ctx := context.Background()
+	chunksCh := make(chan *sources.Chunk, 1)
+	source := &Source{
+		gcsManager: &mockObjectManager{},
+		chunksCh:   chunksCh,
+		Progress:   sources.Progress{},
+	}
+
+	cache := memory.New()
+	source.cache = newPersistableCache(5, cache, &source.Progress)
+
+	err := source.enumerate(ctx)
+	assert.Nil(t, err)
+
+	go func() {
+		defer close(chunksCh)
+		err := source.Chunks(ctx, chunksCh)
+		assert.Nil(t, err)
+	}()
+
+	want := make([]*sources.Chunk, 0, 5)
+	for i := 0; i < 5; i++ {
+		want = append(want, createTestSourceChunk(i))
+	}
+
+	got := make([]*sources.Chunk, 0, 5)
+	for ch := range chunksCh {
+		got = append(got, ch)
+	}
+
+	// Ensure we get 5 objects back.
+	assert.Equal(t, len(want), len(got))
+
+	processing := map[string]struct{}{"object0": {}, "object1": {}, "object2": {}, "object3": {}, "object4": {}}
+	for i := 0; i < 5; i++ {
+		delete(processing, fmt.Sprintf("object%d", i))
+	}
+
+	// Test that the resume progress is set.
+	// The processed values should be the greatest object name lexicographically.
+	progress := "object0,object1,object2,object3,object4"
+	objs := strings.Split(progress, ",")
+
+	encodeResume := strings.Split(source.Progress.EncodedResumeInfo, ",")
+	sort.Strings(encodeResume)
+
+	assert.Equal(t, objs, encodeResume)
+	assert.Equal(t, int32(5), source.Progress.SectionsCompleted)
+	assert.Equal(t, int64(100), source.Progress.PercentComplete)
+	assert.Equal(t, 0, source.cache.Count())
```

Comment:
Ah, I see it now. I asked "why is the cache cleared" but I probably should have asked "why _was_ the cache cleared," not because it _shouldn't_ be cleared, but because once the run is complete it feels like it shouldn't exist at all! But it has to exist post-run, because it's stored in `Source`, because that's how you persist information from `Init` through to `Chunks`. That implementation constraint forces the cache to be present here, so we might as well manage its state, and verify our management. Got it.

But! As a **suggestion**, could you set the cache up inside `Chunks` instead of `Init`? (It looks like you can as a matter of simply moving the code around, but I'm not familiar with implicit assumptions about those functions, so maybe there's a reason it won't work.) If so, you don't need to store the cache (or its mutex or extra count variable) inside `Source`, so you don't have to worry about clearing or checking it after the run.

Reply:
I guess one reason why the cache might make sense to still be a part of the source and get initialized in Init would be when we transition to distributed jobs. We would have the job split across instances and they would all need to reference the same cache for progress tracking.

That being said, I might be thinking too far in advance, and we can cross that bridge when we get there.